### DISCUSSION TITLE
How to make changes to the controller using the event system in version 4.0.2.3?

### DISCUSSION THREAD

**Discussion by imtaxu** [opencart/opencart#14221] [Q&A]
I have a custom module that I use for OpenCart 3.0.3.8. In this module, I create and use a custom PHP class under the system/library folder via OCMOD. Additionally, I make changes in the controller/product/product.php files on both the admin and catalog sides. I also modify the system/library/cart/cart.php file, and my module works without any issues.

Now, I am trying to update this module to work with the OpenCart 4.0.2.3 version using the event system. I create an event in the Opencart\Admin\Controller\Extension\CanvasPro\Module file as follows:

```
$productController = [
    'code' => 'canvaspro_catalog_product_controller',
    'description' => 'Canvas Pro Product Controller',
    'trigger' => 'catalog/controller/product/product/*',
    'action' => 'extension/canvaspro/module/canvaspro.productController',
    'status' => '1',
    'sort_order' => '0'
];

$this->model_setting_event->addEvent($productController);
```

Then, I define the productController function in the Opencart\Catalog\Controller\Extension\CanvasPro\Module file:

```
public function productController(string $route, array $data, string|null &$output): void
{
    echo "<pre>";
    echo $route;
    echo '<br>';
    echo 'data: ';
    print_r($data);
    echo '<br>';
    echo 'output';
    print_r($output);
    echo "</pre>";
    
    // $output = str_replace($data['heading_title'], $data['heading_title'] . ' Extended Output', $output);

    exit();
}
```

I have tried using before, after, and * values in the trigger section. I also tested sort_order values between -99 and +99. The event is successfully registered in the system. The $route value is displayed as product/product. The $data value appears as an empty array, and $output is returned as null.

However, I cannot overwrite any values or inject any code into the catalog/controller/product/product.php file. For example, I want to manipulate the $data['heading_title'] value or add my custom code right before or after the corresponding line, but I haven't been able to achieve this.

Could you please point out what I am doing wrong? Thank you.

**Comment by osworx** [comment]
You should better call such things in the OpenCart Forum: https://forum.opencart.com/ 
Because your question is not related to the core here.

Beside this, you have not stated if this is an event **before** or **after** (see trigger) - which is very important what and how you will get the data.

**Comment by mhcwebdesign** [comment]
If all you want to do is the modify the heading_title, then you should write an event handler for the trigger **catalog/view/product/product/before** .

Please post on the OpenCart forum to get some more help.

**Comment by imtaxu** [comment]
The reason I'm posting here on GitHub discussions instead of the OpenCart forum is to get detailed information from developers about the usage of before/after events.

All the articles I've read about the event system explain the topic superficially. I can make the modifications I want in view files, and the events work fine. Although I haven't tested it yet, as far as I understand, methods can be added and modified in model files as well.
However, in controller files, I cannot manipulate any data using before or after events.

For example, by default, "+" or "-" prefix values can be used for prices in product options. I want to add the "=" prefix value to enter fixed prices for options. I've made the modifications I want on the admin side. On the catalog side, to add to cart according to the product option price, I want to modify this line in catalog/controller/product/product.php:

```
$price = $this->currency->format($this->tax->calculate($option_value['price'], $product_info['tax_class_id'], $this->config->get('config_tax') ? 'P' : false);
```

to this:

```
if ($option_value['price_prefix']=="u") {
    $price = '+' . (float)$option_value['price'].'%';
}
elseif ($option_value['price_prefix']=="d") {
    $price = '-' . (float)$option_value['price'].'%';
}
else{
    $price = $option_value['price_prefix'].$this->currency->format($this->tax->calculate($option_value['price'], $product_info['tax_class_id'], $this->config->get('config_tax') ? 'P' : false), $this->session->data['currency']);
}
```

I've been working on this for 4 days and tried everything, but unfortunately, my efforts haven't yielded results.
Since I couldn't find detailed information about the before/after usage of the event system for controller files, I'm asking for help from you, valued masters.
Thanks.

**Comment by imtaxu** [comment]
```
$this->model_setting_event->addEvent([
            'code' => 'system_library_cart_extend',
            'description' => '',
            'trigger' => 'library/cart/cart/before',
            'action' => 'extension/mymodule/system/cart.systemCart',
            'status' => '1',
            'sort_order' => '1'
]);
```

extension/myodule/system/cart.php

```
<?php

namespace Opencart\Extension\CanvasPro\System\Cart;

class Cart extends \Opencart\System\Library\Cart
{
    public function systemCart (string &$route, array &$args, string &$template = ''): void
    {
        echo $route;
        exit;
    }
}
```

for example this event also does not work. I try to update $option_price = 0; to $option_price = array(); but it doesn't work. 

i guess it makes more sense to switch back to vqmod in this case.

**Comment by mhcwebdesign** [comment]
The **system/library/cart.php** isn't directly supported by event handlers. I usually create a decorator wrapping the cart instance, and then use a controller/*/before trigger to update the registry, e.g. $this->registry->set($myCartDecorator).
